src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts

```diff
@@ -203,7 +228,15 @@ export class IndexPattern implements IIndexPattern {
     this.timeFieldName = spec.timeFieldName;
     this.sourceFilters = spec.sourceFilters;
 
-    this.fields.replaceAll(spec.fields || []);
+    try {
+      this.fields.replaceAll(spec.fields || []);
+    } catch (err) {
+      if (err instanceof FieldTypeUnknownError) {
+        this.unknownFieldErrorNotification(err.fieldSpec.name, err.fieldSpec.type, this.title);
+      } else {
+        throw err;
+      }
+    }
     this.typeMeta = spec.typeMeta;
 
     this.fieldFormatMap = _.mapValues(fieldFormatMap, (mapping) => {
```

Comment:
lets add the format to the `Field` (non deserialized, `SerializedFieldFormat`)

Reply:
and lets get rid of the fieldFormatMap

Reply:
Luckily the field formatter registry already memoizes the creation of field formatter instances so I don't think we need to worry about performance concerns.

Reply:
We can't really get rid of the fieldFormatMap because thats how field formats are persisted. Because we wish to load the field list dynamically, we need to save the fieldFormat info separately from the field list. There are also cases where we'd like to be able to save fieldFormat info without the field existing. 

Can you explain the desire to provide a serialized fieldFormat on the `IndexPatternField` class? Strikes me as an odd mix of serialized and nonserialized code.

The additional implication of providing a FieldFormat instance on a IndexPatternField instance is that the creator would need access to the FieldFormatsRegistry. Currently the index pattern has a reference. Perhaps we need `indexPattern.addField`.

Reply:
`indexPatternField.format` is the most referenced bit of index pattern related formatting code. Its referenced by esQuery, aggs, discover, visualizations, maps, and lens. I'd like to make sure we do a good cost benefit analysis before changing it. I'd prefer to leaving format related changes alone for this PR if possible. I'm not sure where that leaves passing `getFormatterForField` - that could be determined by the code consuming it.

Reply:
> lets add the format to the Field (non deserialized, SerializedFieldFormat)

I like this idea of having the `SerializedFieldFormat` stored on the index pattern field directly, (but not the instance itself). IMO this creates a nice separation of concerns for downstream users; if you need the field format, you can get the serialized version from the index pattern, but must go to the field formats service and use `fieldFormats.deserialize` if you want the actual formatter instance.

It doesn't change the dependencies of index patterns internally (you still need access to the field formats registry), but it forces external users to be explicit about whether or not they rely on the field formatters.

```js
const formatter = fieldFormats.deserialize(indexPattern.getFieldByName('myField').format);
```

The current implementation also feels a little awkward in scenarios where you only want a serialized format -- for example the changes that were made to aggs now look like this:

```js
      const format = agg.params.field
        ? agg.aggConfigs.indexPattern.getFormatterForField(agg.params.field).toJSON()
        : { id: undefined, params: undefined };
```

Whereas if you provided the serialized format, it would be simpler:

```js
      const format = agg.aggConfigs.indexPattern.getFieldByName(agg.params.field.name).format
        ?? { id: undefined, params: undefined };
```

Providing the serialized formats this way would also remove the need to pass `getFormatterForField` into that field `toSpec` method, which feels somewhat odd to me too (I would assume that if I have an instance which I need to serialize, I don't need to provide more dependencies to it in order to perform the serialization).

An approach like this is not without precedent -- it is how we handle things inside agg config at the moment. There's an `aggConfig.toSerializedFieldFormat()` which will return `SerializedFieldFormat` for the field you are aggregating on. This was done intentionally so that you must use `fieldFormats.deserialize` if you need the formatter instance, rather than getting that instance from a different service.

Reply:
>but it forces external users to be explicit about whether or not they rely on the field formatters.

I think this specifically is worth resolving. Is there value in forcing users to be explicit about dependencies? Consider it in abstract - one situation you grab the IndexPattern service which includes field formatters, another you need to grab the Index Pattern service and the field formatter service. Which is better? 

As best I understand, the cases where IndexPatterns is used without field formatting is pretty small. You'd be querying for documents but not formatting the result. 

In the case where the dev needs two dependencies - how are they benefitting from this additional work? I'm having trouble finding a benefit. 

In the case where a dev doesn't need formatting - is there something undesirable about its presence? I could see it affecting bundle size, but I think this is a premature optimization unless we back it up with numbers.

Side note - the IndexPattern class currently handles looking up default formatters. Needs to be accounted for but should be simple to move. Actually, the reason why the `getFormatterForField` is there is to get the default field formatter. Maybe this would be better addressed where the `SerializedFieldFormat` is being consumed rather than supplying it to all `FieldSpec`s